packages/tests-unit/tests/core/routing/util.test.ts

```diff
@@ -0,0 +1,663 @@
+    expect(convertToQuery(querystring)).toEqual({
+      key: ["value1", "value2"],
+      another: "value3",
+    });
+  });
+});
+
+describe("getMiddlewareMatch", () => {
+  it("returns an empty list when root route not matched", () => {
+    expect(getMiddlewareMatch({ middleware: {} })).toEqual([]);
+  });
+
+  it("returns an empty list when root matchers are empty", () => {
+    expect(
+      getMiddlewareMatch({
+        middleware: {
+          "/": {
+            matchers: [],
+          },
+        },
+      }),
+    ).toEqual([]);
+  });
+
+  it("returns a list of regular expressions for each matcher on the root", () => {
+    expect(
+      getMiddlewareMatch({
+        middleware: {
+          "/": {
+            matchers: [{ regexp: "/" }, { regexp: "/path" }],
+          },
+        },
+      }),
+    ).toEqual([/\//, /\/path/]);
+  });
+});
+
+describe("regex", () => {
+  describe("escapeRegex", () => {
+    it("should escape (.) with _µ1_", () => {
+      const result = escapeRegex("/a(.)b");
+      expect(result).toBe("/a_µ1_b");
+    });
+
+    it("should escape (..) with _µ2_", () => {
+      const result = escapeRegex("/a(..)b");
+      expect(result).toBe("/a_µ2_b");
+    });
+
+    it("should escape (...) with _µ3_", () => {
+      const result = escapeRegex("/a(...)b");
+      expect(result).toBe("/a_µ3_b");
+    });
+  });
+
+  describe("unescapeRegex", () => {
+    it("should unescape _µ1_ with (.)", () => {
+      const result = unescapeRegex("/a_µ1_b");
+      expect(result).toBe("/a(.)b");
+    });
+
+    it("should unescape _µ2_ with (..)", () => {
+      const result = unescapeRegex("/a_µ2_b");
+      expect(result).toBe("/a(..)b");
+    });
+
+    it("should unescape _µ3_ with (...)", () => {
+      const result = unescapeRegex("/a_µ3_b");
+      expect(result).toBe("/a(...)b");
+    });
+  });
+
+  it.each([
+    "/a(.)b",
+    "/a(..)b",
+    "/a(...)b",
+    "/a(....)b",
+    "/some/path",
+    "/some/file.json",
+  ])("should escape and unescape %s", (input) => {
+    const escaped = escapeRegex(input);
+    const unescaped = unescapeRegex(escaped);
+    expect(unescaped).toBe(input);
+  });
+});
+
+describe("convertBodyToReadableStream", () => {
+  it("returns undefined for GET requests", () => {
+    const result = convertBodyToReadableStream("GET");
+    expect(result).toEqual(undefined);
+  });
+
+  it("returns undefined for HEAD requests", () => {
+    const result = convertBodyToReadableStream("HEAD");
+    expect(result).toEqual(undefined);
+  });
+
+  it("returns undefined when body is undefined", () => {
+    const result = convertBodyToReadableStream("POST");
+    expect(result).toEqual(undefined);
+  });
+
+  it("returns readable stream for when body is provided", async () => {
+    const result = convertBodyToReadableStream("PUT", Buffer.from("body"));
+    expect(await fromReadableStream(result as any)).toEqual("body");
+  });
+});
+
+describe.skip("TODO: proxyRequest", () => {});
```

Comment:
Yeah, I thought it would be difficult too.

Did you want me to remove this altogether or update the description to say that it covered by e2e tests?

Reply:
Updating the description makes more sense so that if someone change they'll know that they should implement test in e2e